# **Решение задач по библиотеке Pandas**

# **1. Подготовка исходных данных**

Считать файл 'transactions.csv' и записать его в таблицу transactions.

Таблица содержит следующие признаки:
- product_id - идентификатор продукта;
- customer_id	- идентификатор клиента;
- store_id - идентификатор магазина;
- month_of_year	- месяц покупки;
- store_sales - стоимость покупки.

In [2]:
import pandas as pd
import numpy as np

tr = pd.read_csv("transactions.csv")
tr.head(10)

product_id  customer_id  store_id  month_of_year  store_sales
0           1          157        24             12         8.55
1           1          456        15              6        11.40
2           1          638        11              9         8.55
3           1          916         7              4        11.40
4           1          923        15              7         8.55
5           1         1312         3              5         8.55
6           1         1565        24              9         8.55
7           1         2270        11             11         8.55
8           1         3065         3             11         5.70
9           1         3441         3              8         8.55

Считать файл 'products.csv' и записать его в таблицу products.

Таблица содержит следующие признаки:
- product_class_id - идентификатор категории продукта;
- product_id - идентификатор продукта;
- brand_name - производитель продукта;
- product_name - название продукта.

In [3]:
pr = pd.read_csv("products.csv")
pr.head(10)

product_class_id  product_id  brand_name                 product_name
0                30           1  Washington       Washington Berry Juice
1                52           2  Washington       Washington Mango Drink
2                52           3  Washington  Washington Strawberry Drink
3                19           4  Washington        Washington Cream Soda
4                19           5  Washington         Washington Diet Soda
5                19           6  Washington              Washington Cola
6                19           7  Washington         Washington Diet Cola
7                30           8  Washington      Washington Orange Juice
8                30           9  Washington   Washington Cranberry Juice
9                30          10  Washington       Washington Apple Juice

Считать файл 'product_class.csv' и записать его в таблицу product_class.

Таблица содержит следующие признаки:
- product_class_id - идентификатор категории продукта;
- product_subcategory - название подкатегории продукта;
- product_category - название категории продукта;
- product_department - отдел магазина;
- product_family - продуктовые или продовольственные продукты.

In [4]:
prcl = pd.read_csv("product_class.csv")
prcl.head(10)

product_class_id product_subcategory product_category product_department  \
0                 1                Nuts        Specialty            Produce   
1                 2           Shellfish          Seafood            Seafood   
2                 3        Canned Fruit            Fruit    Canned Products   
3                 4              Spices     Baking Goods       Baking Goods   
4                 5               Pasta    Starchy Foods      Starchy Foods   
5                 6              Yogurt            Dairy              Dairy   
6                 7              Coffee        Dry Goods       Baking Goods   
7                 8          Deli Meats             Meat               Deli   
8                 9           Ice Cream  Frozen Desserts       Frozen Foods   
9                10           TV Dinner   Frozen Entrees       Frozen Foods   

  product_family  
0           Food  
1           Food  
2           Food  
3           Food  
4           Food  
5           Food  
6          Drink  
7           Food  
8           Food  
9           Food

# **2. Объединение таблиц**

На основе таблицы 'transactions' создать три таблицы: trans1 содержит первые 5000 строк, trans2 - строки с 5000 по 150000, trans3 - с 150000 и до конца таблицы.

In [5]:
trans1, trans2, trans3 = tr.loc[:4999], tr.loc[5000:149999].reset_index(), tr.loc[150000:].reset_index()
trans1.head(10)

product_id  customer_id  store_id  month_of_year  store_sales
0           1          157        24             12         8.55
1           1          456        15              6        11.40
2           1          638        11              9         8.55
3           1          916         7              4        11.40
4           1          923        15              7         8.55
5           1         1312         3              5         8.55
6           1         1565        24              9         8.55
7           1         2270        11             11         8.55
8           1         3065         3             11         5.70
9           1         3441         3              8         8.55

In [6]:
trans2

index  product_id  customer_id  store_id  month_of_year  store_sales
0         5000          93         9747        13              5         1.92
1         5001          93         9863        13             12         2.88
2         5002          93         9928        16             11         4.80
3         5003          93         9960         7              1         2.88
4         5004          93        10186        13              6         3.84
...        ...         ...          ...       ...            ...          ...
144995  149995        1136            0      3239              3        14.44
144996  149996        1136            0      3271              2        14.44
144997  149997        1136            0      3499              1        10.83
144998  149998        1136         1223      4391              6        10.83
144999  149999        1136            0      4498             11        14.44

[145000 rows x 6 columns]

In [7]:
trans3

index  product_id  customer_id  store_id  month_of_year  store_sales
0      150000        1136            0      4693              7        10.83
1      150001        1136          980      4961              3        14.44
2      150002        1136            0      5826             11        10.83
3      150003        1136            0      6061              1         7.22
4      150004        1136          205      6098             10         7.22
...       ...         ...          ...       ...            ...          ...
23597  173597        1559            0      8653              2         3.24
23598  173598        1559            0      8659             12         6.48
23599  173599        1559            0      9271              5         6.48
23600  173600        1559            0      9318              9         4.86
23601  173601        1559            0      9851              2         4.86

[23602 rows x 6 columns]

Объединить таблицы 'trans1' и 'trans2' с использованием метода `concat()`.

In [8]:
conc = pd.concat([trans1, trans2])
conc

product_id  customer_id  store_id  month_of_year  store_sales  \
0                1          157        24             12         8.55   
1                1          456        15              6        11.40   
2                1          638        11              9         8.55   
3                1          916         7              4        11.40   
4                1          923        15              7         8.55   
...            ...          ...       ...            ...          ...   
144995        1136            0      3239              3        14.44   
144996        1136            0      3271              2        14.44   
144997        1136            0      3499              1        10.83   
144998        1136         1223      4391              6        10.83   
144999        1136            0      4498             11        14.44   

           index  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
144995  149995.0  
144996  149996.0  
144997  149997.0  
144998  149998.0  
144999  149999.0  

[150000 rows x 6 columns]

Объединить таблицы 'trans1', 'trans2' и 'trans3' с использованием метода `concat()`.

In [9]:
trans123 = pd.concat([trans1, trans2, trans3])
trans123

product_id  customer_id  store_id  month_of_year  store_sales     index
0               1          157        24             12         8.55       NaN
1               1          456        15              6        11.40       NaN
2               1          638        11              9         8.55       NaN
3               1          916         7              4        11.40       NaN
4               1          923        15              7         8.55       NaN
...           ...          ...       ...            ...          ...       ...
23597        1559            0      8653              2         3.24  173597.0
23598        1559            0      8659             12         6.48  173598.0
23599        1559            0      9271              5         6.48  173599.0
23600        1559            0      9318              9         4.86  173600.0
23601        1559            0      9851              2         4.86  173601.0

[173602 rows x 6 columns]

Добавить в таблицу 'transactions' название продукта с использованием метода `concat()`.

Написать вывод о возможности применения метода `concat()` для решения задачи.

In [10]:
pd.concat([tr, pr["product_name"]], axis=1)

''' 
Задача не решена
concat добавляет определенное количество строк. Когда строки в добавляемой таблице заканчиваются, оставляет занчения NaN. Таким образом значения в 
добавленном стобце не зависят от product_id
'''

' \nЗадача не решена\nconcat добавляет определенное количество строк. Когда строки в добавляемой таблице заканчиваются, оставляет занчения NaN. Таким образом значения в \nдобавленном стобце не зависят от product_id\n'

Добавить в таблицу 'transactions' название продукта с использованием метода `merge()`.

Какой тип объединения необходимо использовать для решения задачи? Объяснить сделанный выбор.

In [11]:
cols = pr[['product_id', 'product_name']]

merged = tr.merge(cols, on='product_id', how='left')
merged.head(10)

product_id  customer_id  store_id  month_of_year  store_sales  \
0           1          157        24             12         8.55   
1           1          456        15              6        11.40   
2           1          638        11              9         8.55   
3           1          916         7              4        11.40   
4           1          923        15              7         8.55   
5           1         1312         3              5         8.55   
6           1         1565        24              9         8.55   
7           1         2270        11             11         8.55   
8           1         3065         3             11         5.70   
9           1         3441         3              8         8.55   

             product_name  
0  Washington Berry Juice  
1  Washington Berry Juice  
2  Washington Berry Juice  
3  Washington Berry Juice  
4  Washington Berry Juice  
5  Washington Berry Juice  
6  Washington Berry Juice  
7  Washington Berry Juice  
8  Washington Berry Juice  
9  Washington Berry Juice

Вывести размер и первые 2 строки таблиц 'products' и 'product_class'.

In [12]:
print(pr.shape, prcl.shape)

(1560, 4) (109, 5)


In [13]:
pr.head(2)

product_class_id  product_id  brand_name            product_name
0                30           1  Washington  Washington Berry Juice
1                52           2  Washington  Washington Mango Drink

In [14]:
prcl.head(2)

product_class_id product_subcategory product_category product_department  \
0                 1                Nuts        Specialty            Produce   
1                 2           Shellfish          Seafood            Seafood   

  product_family  
0           Food  
1           Food

Добавить в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'inner'`. Результат записать в таблицу 'm1'.


In [15]:
pspc = prcl[['product_class_id', 'product_subcategory', 'product_category']]

m1 = pr.merge(pspc, on='product_class_id', how='inner')
m1

product_class_id  product_id  brand_name  \
0                   30           1  Washington   
1                   52           2  Washington   
2                   52           3  Washington   
3                   19           4  Washington   
4                   19           5  Washington   
...                ...         ...         ...   
1550                53        1556         CDR   
1551                33        1557         CDR   
1552                53        1558         CDR   
1553                33        1559         CDR   
1554                31        1560         CDR   

                        product_name product_subcategory      product_category  
0             Washington Berry Juice               Juice  Pure Juice Beverages  
1             Washington Mango Drink     Flavored Drinks                Drinks  
2        Washington Strawberry Drink     Flavored Drinks                Drinks  
3              Washington Cream Soda                Soda  Carbonated Beverages  
4               Washington Diet Soda                Soda  Carbonated Beverages  
...                              ...                 ...                   ...  
1550        CDR Creamy Peanut Butter       Peanut Butter      Jams and Jellies  
1551        CDR Strawberry Preserves           Preserves      Jams and Jellies  
1552  CDR Extra Chunky Peanut Butter       Peanut Butter      Jams and Jellies  
1553             CDR Apple Preserves           Preserves      Jams and Jellies  
1554                 CDR Grape Jelly               Jelly      Jams and Jellies  

[1555 rows x 6 columns]

Найти количество пропущенных значений для каждого признака в результирующей таблице 'm1'.

In [16]:
m1.isna().sum()

product_class_id       0
product_id             0
brand_name             0
product_name           0
product_subcategory    0
product_category       0
dtype: int64

Почему в результирующей таблице 'm1' количество записей уменьшилось с 1560 до 1555?

### **Ответ:** Потому что inner добавляет в результирующую таблицу только те строки, в которых нашлось соответсвие в двух таблицах одновременно

Добавить в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'left'`. Записать полученный результат в таблицу 'm2'.


In [17]:
m2 = pr.merge(pspc, on='product_class_id', how='left')
m2

product_class_id  product_id  brand_name  \
0                   30           1  Washington   
1                   52           2  Washington   
2                   52           3  Washington   
3                   19           4  Washington   
4                   19           5  Washington   
...                ...         ...         ...   
1555                53        1556         CDR   
1556                33        1557         CDR   
1557                53        1558         CDR   
1558                33        1559         CDR   
1559                31        1560         CDR   

                        product_name product_subcategory      product_category  
0             Washington Berry Juice               Juice  Pure Juice Beverages  
1             Washington Mango Drink     Flavored Drinks                Drinks  
2        Washington Strawberry Drink     Flavored Drinks                Drinks  
3              Washington Cream Soda                Soda  Carbonated Beverages  
4               Washington Diet Soda                Soda  Carbonated Beverages  
...                              ...                 ...                   ...  
1555        CDR Creamy Peanut Butter       Peanut Butter      Jams and Jellies  
1556        CDR Strawberry Preserves           Preserves      Jams and Jellies  
1557  CDR Extra Chunky Peanut Butter       Peanut Butter      Jams and Jellies  
1558             CDR Apple Preserves           Preserves      Jams and Jellies  
1559                 CDR Grape Jelly               Jelly      Jams and Jellies  

[1560 rows x 6 columns]

Найти количество пропущенных значений для каждого признака результирующей таблицы 'm2'.

In [18]:
m2.isna().sum()

product_class_id       0
product_id             0
brand_name             0
product_name           0
product_subcategory    5
product_category       5
dtype: int64

Вывести записи таблицы 'm2', в которых есть пропущенные значения.

In [19]:
m2[m2.isna().any(axis=1)]

product_class_id  product_id    brand_name  \
158                 86         159         Denny   
469                 86         470      Red Wing   
780                 86         781     Cormorant   
1096                86        1097  High Quality   
1407                86        1408        Sunset   

                          product_name product_subcategory product_category  
158          Denny Copper Pot Scrubber                 NaN              NaN  
469       Red Wing Copper Pot Scrubber                 NaN              NaN  
780      Cormorant Copper Pot Scrubber                 NaN              NaN  
1096  High Quality Copper Pot Scrubber                 NaN              NaN  
1407        Sunset Copper Pot Scrubber                 NaN              NaN

Сделать вывод о причинах пропусков в результирующей таблице 'm2'.

### **Вывод:** Если при left join в правой таблице не находится соединения для левой таблицы, в результирующую таблицу записывается NaN

Добавить в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'right'`. Записать полученный результат в таблицу 'm3'.


In [20]:
m3 = pr.merge(pspc, on='product_class_id', how='right')
m3

product_class_id  product_id  brand_name                 product_name  \
0                    1       195.0    High Top             High Top Almonds   
1                    1       205.0    High Top          High Top Mixed Nuts   
2                    1       206.0    High Top      High Top Canned Peanuts   
3                    1       211.0    High Top             High Top Walnuts   
4                    1       214.0    High Top          High Top Party Nuts   
...                ...         ...         ...                          ...   
1558               110      1050.0     PigTail      PigTail Lemon Popsicles   
1559               110      1332.0  Carrington  Carrington Orange Popsicles   
1560               110      1349.0  Carrington   Carrington Grape Popsicles   
1561               110      1351.0  Carrington    Carrington Lime Popsicles   
1562               110      1361.0  Carrington   Carrington Lemon Popsicles   

     product_subcategory product_category  
0                   Nuts        Specialty  
1                   Nuts        Specialty  
2                   Nuts        Specialty  
3                   Nuts        Specialty  
4                   Nuts        Specialty  
...                  ...              ...  
1558           Popsicles  Frozen Desserts  
1559           Popsicles  Frozen Desserts  
1560           Popsicles  Frozen Desserts  
1561           Popsicles  Frozen Desserts  
1562           Popsicles  Frozen Desserts  

[1563 rows x 6 columns]

Найти количество пропущенных значений для каждого признака результирующей таблицы 'm3'.

In [21]:
m3.isna().sum()

product_class_id       0
product_id             8
brand_name             8
product_name           8
product_subcategory    0
product_category       0
dtype: int64

Вывести записи результирующей таблицы 'm3', в которых есть пропущенные значения.

In [22]:
m3[m3['brand_name'].isna()]

product_class_id  product_id brand_name product_name  \
83                   7         NaN        NaN          NaN   
204                 13         NaN        NaN          NaN   
300                 21         NaN        NaN          NaN   
316                 23         NaN        NaN          NaN   
829                 59         NaN        NaN          NaN   
1080                68         NaN        NaN          NaN   
1081                69         NaN        NaN          NaN   
1322                92         NaN        NaN          NaN   

     product_subcategory   product_category  
83                Coffee          Dry Goods  
204     Fresh Vegetables         Vegetables  
300             Cleaners  Cleaning Supplies  
316          Shower Soap  Cleaning Supplies  
829      Dehydrated Soup      Packaged Soup  
1080           Dish Soap  Bathroom Products  
1081     Dishwasher Soap  Bathroom Products  
1322  Non-Alcoholic Wine      Beer and Wine

Сделать вывод о причинах пропущенных значений.

### **Вывод:** правой таблице не нашлось соответсвий из левой таблицы, из-за чего при right соединении значения из левой таблицы были пропущены 

Добавить в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'outer'`. Записать полученный результат в таблицу 'm4'.


In [23]:
m4 = pr.merge(pspc, on='product_class_id', how='outer')
m4

product_class_id  product_id  brand_name                 product_name  \
0                    1       195.0    High Top             High Top Almonds   
1                    1       205.0    High Top          High Top Mixed Nuts   
2                    1       206.0    High Top      High Top Canned Peanuts   
3                    1       211.0    High Top             High Top Walnuts   
4                    1       214.0    High Top          High Top Party Nuts   
...                ...         ...         ...                          ...   
1563               110      1050.0     PigTail      PigTail Lemon Popsicles   
1564               110      1332.0  Carrington  Carrington Orange Popsicles   
1565               110      1349.0  Carrington   Carrington Grape Popsicles   
1566               110      1351.0  Carrington    Carrington Lime Popsicles   
1567               110      1361.0  Carrington   Carrington Lemon Popsicles   

     product_subcategory product_category  
0                   Nuts        Specialty  
1                   Nuts        Specialty  
2                   Nuts        Specialty  
3                   Nuts        Specialty  
4                   Nuts        Specialty  
...                  ...              ...  
1563           Popsicles  Frozen Desserts  
1564           Popsicles  Frozen Desserts  
1565           Popsicles  Frozen Desserts  
1566           Popsicles  Frozen Desserts  
1567           Popsicles  Frozen Desserts  

[1568 rows x 6 columns]

Найти количество пропущенных значений для каждого признака результирующей таблицы 'm4'.

In [24]:
m4.isna().sum()

product_class_id       0
product_id             8
brand_name             8
product_name           8
product_subcategory    5
product_category       5
dtype: int64

Вывести записи результирующей таблицы 'm4', в которых есть пропущенные значения.

In [25]:
m4[m4.isna().any(axis=1)]

product_class_id  product_id    brand_name  \
83                   7         NaN           NaN   
204                 13         NaN           NaN   
300                 21         NaN           NaN   
316                 23         NaN           NaN   
829                 59         NaN           NaN   
1080                68         NaN           NaN   
1081                69         NaN           NaN   
1267                86       159.0         Denny   
1268                86       470.0      Red Wing   
1269                86       781.0     Cormorant   
1270                86      1097.0  High Quality   
1271                86      1408.0        Sunset   
1327                92         NaN           NaN   

                          product_name product_subcategory   product_category  
83                                 NaN              Coffee          Dry Goods  
204                                NaN    Fresh Vegetables         Vegetables  
300                                NaN            Cleaners  Cleaning Supplies  
316                                NaN         Shower Soap  Cleaning Supplies  
829                                NaN     Dehydrated Soup      Packaged Soup  
1080                               NaN           Dish Soap  Bathroom Products  
1081                               NaN     Dishwasher Soap  Bathroom Products  
1267         Denny Copper Pot Scrubber                 NaN                NaN  
1268      Red Wing Copper Pot Scrubber                 NaN                NaN  
1269     Cormorant Copper Pot Scrubber                 NaN                NaN  
1270  High Quality Copper Pot Scrubber                 NaN                NaN  
1271        Sunset Copper Pot Scrubber                 NaN                NaN  
1327                               NaN  Non-Alcoholic Wine      Beer and Wine

Сделать вывод о причинах пропущенных значений.

### **Вывод:** Все значения левой таблицы не имеющие соответвия в правой таблице и все значения правой таблицы не имеющие соответсвия в левой таблице обрели значения NaN. Можно сравнить с left join + right join

В таблицу 'transactions' добавить категорию каждого проданного продукта.

In [26]:
tr.merge(pr[['product_id', 'product_class_id']].merge(prcl[['product_class_id', 'product_category']], on='product_class_id', how='inner')[['product_id', 'product_category']], on='product_id', how='inner').head(10)

product_id  customer_id  store_id  month_of_year  store_sales  \
0           1          157        24             12         8.55   
1           1          456        15              6        11.40   
2           1          638        11              9         8.55   
3           1          916         7              4        11.40   
4           1          923        15              7         8.55   
5           1         1312         3              5         8.55   
6           1         1565        24              9         8.55   
7           1         2270        11             11         8.55   
8           1         3065         3             11         5.70   
9           1         3441         3              8         8.55   

       product_category  
0  Pure Juice Beverages  
1  Pure Juice Beverages  
2  Pure Juice Beverages  
3  Pure Juice Beverages  
4  Pure Juice Beverages  
5  Pure Juice Beverages  
6  Pure Juice Beverages  
7  Pure Juice Beverages  
8  Pure Juice Beverages  
9  Pure Juice Beverages

# **3. Агрегирование и группировка данных**

Объеденить таблицы 'transactions' и 'products' с использованием метода `merge()`.

In [27]:
tr.merge(pr)

product_id  customer_id  store_id  month_of_year  store_sales  \
0                1          157        24             12         8.55   
1                1          456        15              6        11.40   
2                1          638        11              9         8.55   
3                1          916         7              4        11.40   
4                1          923        15              7         8.55   
...            ...          ...       ...            ...          ...   
173597        1559            0      8653              2         3.24   
173598        1559            0      8659             12         6.48   
173599        1559            0      9271              5         6.48   
173600        1559            0      9318              9         4.86   
173601        1559            0      9851              2         4.86   

        product_class_id  brand_name            product_name  
0                     30  Washington  Washington Berry Juice  
1                     30  Washington  Washington Berry Juice  
2                     30  Washington  Washington Berry Juice  
3                     30  Washington  Washington Berry Juice  
4                     30  Washington  Washington Berry Juice  
...                  ...         ...                     ...  
173597                33         CDR     CDR Apple Preserves  
173598                33         CDR     CDR Apple Preserves  
173599                33         CDR     CDR Apple Preserves  
173600                33         CDR     CDR Apple Preserves  
173601                33         CDR     CDR Apple Preserves  

[173602 rows x 8 columns]

Вывести уникальные значения признака 'product_class_id'.

In [28]:
pr["product_class_id"].unique()

array([ 30,  52,  19,  35,  62,  58,  93,  94,  95,  96,  97,  98,   3,
        89, 103, 102,  15,  14,  11,  76,   6,  81,  91,  16,  77,   8,
        66,  34, 110, 100,  29,  63,  64,  60,  10,  28,   9,  78,  39,
        79,  88,  70,  71,  72,  73,  75,  74,  80,  56, 101,  55,  22,
        18,  20,  40,  87,  47,  86,  85,  41,  42,  65, 104, 108, 107,
       105, 106,  61,  99,  67,   1,  24,   2,  45,  12,  82,  54,  83,
        46,  84,  17, 109,  36,  37,  38,   5,  57,  43,  44,  27,  26,
        25,  49,  53,  51,  50,  90,  48,  33,  31,   4,  32], dtype=int64)

Определить количество покупок, приходящихся на каждый месяц.

Определить долю покупок, приходящихся на каждый месяц.

In [29]:
tr[['month_of_year', 'store_sales']].groupby('month_of_year').count()

store_sales
month_of_year             
1                    14064
2                    13682
3                    15414
4                    13170
5                    13726
6                    13818
7                    15502
8                    14074
9                    13316
10                   12950
11                   16460
12                   17426

Вычислить минимальную, максимальную и среднюю стоимть покупок (store_sales).

In [30]:
tr[['store_sales']].agg(['min', 'max', 'mean'])

store_sales
min      0.500000
max     30.560000
mean     6.511619

Вычислить среднюю стоимость покупок для каждого клиента.

In [69]:
m = tr[['customer_id', 'store_sales']].groupby('customer_id').agg(['sum', 'count'])
m['sred'] = m['store_sales']['sum'] / m['store_sales']['count']
m 


store_sales              sred
                    sum  count           
customer_id                              
0             414003.82  63627   6.506732
3                 98.54     17   5.796471
5                  1.08      1   1.080000
6                 65.76      6  10.960000
10                53.82      9   5.980000
...                 ...    ...        ...
10268             22.26      2  11.130000
10273            329.83     48   6.871458
10274            191.09     24   7.962083
10275             44.02      7   6.288571
10277             29.96      5   5.992000

[5653 rows x 3 columns]

Вычислить сумму, среднее значение и количество покупок для каждого клиента.

In [32]:
tr[['customer_id', 'store_sales']].groupby('customer_id').agg(['sum', 'mean', 'count'])

store_sales                  
                    sum       mean  count
customer_id                              
0             414003.82   6.506732  63627
3                 98.54   5.796471     17
5                  1.08   1.080000      1
6                 65.76  10.960000      6
10                53.82   5.980000      9
...                 ...        ...    ...
10268             22.26  11.130000      2
10273            329.83   6.871458     48
10274            191.09   7.962083     24
10275             44.02   6.288571      7
10277             29.96   5.992000      5

[5653 rows x 3 columns]

Сформировать таблицу 'purch_list', которая должна содержать список всех покупок и количество покупок для каждого клиента.

Для расчета количества покупок использовать признак 'store_id'.

Переименовать признаки 'product_name' и 'store_id' в 'purchases' и 'num_purhcases' соответственно.

Отсортировать записи в таблице по возрастанию количества покупок.

In [76]:

purch_list = tr.groupby('customer_id').agg(num_purchases=('store_id', lambda x: list(set(x))), purchases=('store_id', 'count')).sort_values(by='purchases')
purch_list

num_purchases  purchases
customer_id                                                              
8689                                                      [14]          1
3346                                                       [7]          1
2374                                                      [15]          1
4293                                                      [14]          1
2377                                                       [7]          1
...                                                        ...        ...
829          [2053, 4614, 5639, 1547, 8720, 1041, 2039, 923...        457
523          [4100, 6, 523, 527, 4115, 4627, 6696, 7210, 31...        626
980          [2052, 3589, 6, 7, 1035, 1548, 7694, 543, 9760...        731
504          [5120, 3585, 9218, 1538, 8709, 3079, 6665, 13,...        869
0            [3, 6, 10, 14, 19, 20, 23, 26, 28, 30, 36, 39,...      63627

[5653 rows x 2 columns]

Определить сумму покупок для каждого магазина ('store_id').

Признак 'store_id' преобразовать из индекса в столбец.

In [79]:
tr[['store_id', 'store_sales']].groupby('store_id').count()

store_sales
store_id             
2                1380
3                7886
5                   1
6                6821
7                8206
...               ...
10268               2
10273              48
10274              24
10275               7
10277               5

[5589 rows x 1 columns]

Определить среднюю стоимость покупок в различных магазинах ('store_id') для продуктов различных брендов ('brand_name').

In [88]:
tr[['store_id', 'store_sales', 'product_id']].merge(pr[['product_id', 'brand_name']], on='product_id', how='inner')[['store_id', 'store_sales', 'brand_name']].groupby('brand_name').agg(sales_mean=('store_sales', 'mean'), stores=('store_id', lambda x: list(set(x))))

sales_mean                                             stores
brand_name                                                               
ADJ           7.804138  [5376, 1153, 2, 3, 5123, 6, 7, 1159, 2441, 986...
Akron         5.341043  [2, 3, 4100, 6, 7, 11, 13, 15, 16, 17, 5650, 9...
American      6.801710  [4096, 10240, 2, 3, 6148, 2053, 6, 7, 4099, 61...
Amigo         9.016870  [2, 3, 6, 7, 9224, 4618, 11, 2572, 13, 14, 15,...
Applause      7.974138  [2, 3, 4099, 6, 7, 6153, 11, 6155, 13, 14, 15,...
...                ...                                                ...
Toucan        4.293821  [2, 3, 6149, 6, 7, 9224, 4618, 11, 1548, 13, 1...
Tri-State     6.596892  [8193, 2, 3, 8194, 6, 7, 8202, 11, 13, 14, 15,...
Urban         6.770602  [10240, 9728, 2, 3, 4099, 5633, 6, 7, 10243, 6...
Walrus        7.274639  [2, 3, 6, 7, 10250, 11, 13, 14, 15, 16, 17, 22...
Washington    6.771649  [2, 3, 2052, 6, 7, 8202, 11, 6155, 13, 14, 15,...

[111 rows x 2 columns]

Определить среднюю стоимость покупок в различных магазинах ('store_id') для продуктов различных брендов ('brand_name').

Использовать механизм сводных таблиц (метод `pivot_table()`).

Определить количество продаж в каждом месяце ('month_of_year') для каждого бренда ('brand_name').

Использовать кросс-таблицы.

In [37]:
# код

# **4. Операции со строковыми данными**

Считать файл 'imdbratings.csv' и записать его в таблицу 'film'.

In [38]:
# код

Преобразовать названия признаков так, чтобы они начинались с заглавной буквы.

In [39]:
# код

Для возможности сортировки фильмов по названию записать артикль 'The' в конце названия через запятую, например:
- было 'The Shawshank Redemption';
- должно стать 'Shawshank Redemption, The'

In [40]:
# код

Поменять название жанра 'Sci-Fi' на 'Science fiction'.

In [41]:
# код

Вывести информацию о фильмах, в которых снимался Brad Pitt.

In [42]:
# код

# **5. Операции с датой и временем**

Создать объект класса дата/время `Timestamp` вида 16 апреля 2024 года 10 часов 20 минут на основе:
- перечня целых чисел;
- строковой записи даты;
- целого числа, представляющего Unix-время.

In [43]:
# код

Создать объект дата/время `Timestamp` вида 16 апреля 2024 года 10 часов 20 минут с использованием метода `to_datetime()` на основе:
- перечня целых чисел;
- строковой записи даты;
- целого числа, представляющего Unix-время.

In [44]:
# код

Создать объект класса `DatetimeIndex`,  включающий даты начиная с 16 апреля 2024 года по 19 апреля 2024 года включительно с перидичностью 1 день.

Использовать:
- метод `DatetimeIndex()`;
- метод `to_datetime()`;
- метод `date_range()`.

In [45]:
# код

Создать объект `Timedelta` с временным интервалом 1 день.

In [46]:
# код

Вывести с помощью созданного объекта дату следующего дня.

In [47]:
# код

Создать объект `TimedeltaIndex`, содержащий 15 временных дельт в 1 час.

Использовать:
- метод `TimedeltaIndex()`;
- метод `to_timedelta()`;
- метод `timedelta_range()`.

In [48]:
# код

Для выполнения следующих заданий используется набор данных о динамике велосипедного движения между центром города и некоторым его районом.

Считать файл 'time_series.csv' и записать его в таблицу 'df'.

In [49]:
df = # код

SyntaxError: invalid syntax (1403900460.py, line 1)

Объединить столбцы 'Date' и 'Time' в один столбец 'Date_Time'.

Элемент нового столбца должен иметь следующий вид:
14/07/2011	17:44

In [ ]:
# код

Удалить столбцы 'Date' и 'Time'.

In [ ]:
# код

Преобразовать тип переменной 'Date_Time' в формат `datetime`.


In [ ]:
# код

Установить столбец 'Date_Time' в качестве индекса.

In [ ]:
# код

Добавить признак 'both' с суммарным потоком.

In [ ]:
# код

Добавить столбец 'weekday' с названиями дней недели на русском языке.

In [ ]:
# код

Выбрать временные интервалы, для которых суммарный поток менее 3 велосипедистов.

In [ ]:
# код

Выбрать временные интервалы, для которых в одном из направлений проехало менее 5 велосипедистов.

In [ ]:
# код

Определить среднее количество велосипедистов за 15 минут.

In [ ]:
# код

Определить среднее количество велосипедистов за 1 час.

In [ ]:
# код

Найти даты с 5-ю самыми большими значениями количества велосипедистов, проехавших в направлении к центру.

In [ ]:
# код

Найти дату с максимальным количеством велосипедистов за один день в обоих направлениях.

In [ ]:
# код